src/client/gui/gui_cmd.cpp

```diff
@@ -105,6 +106,16 @@ mp::ReturnCode cmd::GuiCmd::run(mp::ArgParser* parser)
     create_menu();
     tray_icon.show();
 
+    QFile first_run_file(QStandardPaths::writableLocation(QStandardPaths::AppDataLocation) + "/first_run");
+
+    if (!first_run_file.exists())
+    {
+        tray_icon.showMessage("Multipass System Tray Menu", "The Multipass system tray menu is now running",
+                              tray_icon.icon());
+        first_run_file.open(QIODevice::WriteOnly);
+        first_run_file.close();
+    }
```

Comment:
I wonder if we should just use the `Settings` class for this to avoid multiplying entities?

Reply:
Hmm, that would work, but it does somewhat expose a key to the user to muck around with.  That said, we wouldn't explicitly expose the key, so the user would have to look at source code to figure out how to set/unset it.  If we're fine with that, then I can move it into the `Settings` handler.

Reply:
Well, it is not a setting, is it? It's an internal flag that we want to persist to and read from disk. The `Settings` class includes that capability but, as Chris points out, it goes beyond it, exposing things to the user. I'd rather not abuse it and leak internal implementation details. 

That would affect future settings frontends, including support for text editing (we'd have to identify and filter things). It is safer and cleaner to keep invariants alone than changing concepts just to include extra entities.

The extra entity will be odd either way. It has less of an impact placed alone than in the wrong group. If such entities got multiplied, a separate group could be created, similar to settings, but with a separate file and no user exposure.

Reply:
I can easily see this as a "hidden" setting, when we introduce more "first run" behaviours, the user may find out that there is a "first_run" hidden setting if they want to go through them again (think a wizard or something).

Reply:
At the very least, we should be able to use the same concept that Settings does for any such internal state storage.

I would rather make this a hidden setting for users to "muck about" with rather than have them muck about with the files directly.

Reply:
One can always find possible uses for an implementation detail. If they come about, it can be consciously promoted to interface.

Settings are part of our interface, internal files aren't. If the user tampers with those, they're on their own. 

I understand the impulse to reuse functionality, I still don't see a strong reason to move the interface/impl boundary, especially since it's 3 LOC above.

Reply:
Thinking it about some more and reading the feedback, it just seems unnatural a "one and done" thing would be a setting.  To me, if we wanted to make this a setting, it should either always show the startup notification or not, something like `client.gui.show-start-notification` and be a boolean.  Otherwise, for true "first run" type of things, we should rely on something else such as what I currently have.

Reply:
> I understand the impulse to reuse functionality, but I still don't see a strong reason to move the interface/impl boundary, especially since it's 3 LOC above.

3 LOC here, 3 in a bunch of other places around the code… Just look in `/var/snap/multipass/common/data/multipassd` at the amount of files we're already storing some state in. Each of those can get corrupted and we have little protection against that. We're heavily multithreaded and it's mostly by chance that we're not stepping on one another's toes when writing / reading them.

Which is why I suggested that the _concept_ behind the Settings class can be useful for this, even if I got outvoted about allowing `multipass set/get …` to affect this state.

> it just seems unnatural a "one and done" thing would be a setting

To me, on the other hand, this is just state. I don't see a difference between a file that is created (and can be removed) and a boolean stored in a file somewhere. It's not like we can blow a fuse :).

That got me thinking, btw, that this shouldn't be a blanket "first run done". We should be more granular - "first run notification shown", because if we introduce some other do-once behaviours, we're likely to want them to run for users that ran Multipass already, but that have not, say, gone through a first-run wizard.

Reply:
> Just look in /var/snap/multipass/common/data/multipassd at the amount of files we're already storing some state in.

So what? Files and dirs are nice units to organize non-volatile info. I am sure one or two things could be improved, but I don't see a structural problem.

> Each of those can get corrupted and we have little protection against that.

Anything can get corrupted. Whether the bytes are in separate or the same file has no bearing on it.

> To me, on the other hand, this is just state. I don't see a difference between a file that is created (and can be removed) and a boolean stored in a file somewhere.

The difference is public vs private &ndash; crucial distinction, probably the 1st consideration to organize data. The direct correspondence between our `Settings` concept and interface data is valuable. It makes everything clean and easy to reason about. It is worth preserving. Lumping everything together would hamper organization, not improve it.

> We should be more granular

I'd favor introducing complexity when it is needed.

Reply:
> To me, on the other hand, this is just state. I don't see a difference between a file that is created (and can be removed) and a boolean stored in a file somewhere.

I definitely understand wanting to keep all persistent gui states in a single file. I do think these states should not be modifiable by the user other than potential interactions with the gui itself, so the keys should not be exposed to the user and not be modifiable by `multipass set`.  

I read up on the `QSettings` documentation and it suggests using it for storing persistent gui states.  So I think in the long run, using our `Settings` infrastructure is the right thing to do, but that requires expanding it to be able handle "hidden" keys, etc.

Given the time crunch of wanting another release out very soon, I don't see that it is possible to implement it in our `Settings` code right now.  we are going to have tech debt for this, so I say we use the simplest way to make it work, like what I currently have.  Either that, or drop this notification for now and add it in when we have time to implement it correctly.

Reply:
> > Just look in /var/snap/multipass/common/data/multipassd at the amount of files we're already storing some state in.
> 
> So what? Files and dirs are nice units to organize non-volatile info. I am sure one or two things could be improved, but I don't see a structural problem.
>
> > Each of those can get corrupted and we have little protection against that.
> 
> Anything can get corrupted. Whether the bytes are in separate or the same file has no bearing on it.

IMO the structural problem is that we don't have a central place that we handle this.
 
> > To me, on the other hand, this is just state. I don't see a difference between a file that is created (and can be removed) and a boolean stored in a file somewhere.
> 
> The difference is public vs private – crucial distinction, probably the 1st consideration to organize data. The direct correspondence between our `Settings` concept and interface data is valuable. It makes everything clean and easy to reason about. It is worth preserving. Lumping everything together would hamper organization, not improve it.

Which is why I said I'd be OK with the _concept_ behind Settings (the singleton class, the storage) being used for state, too - without exposing it to the user.

> > We should be more granular
> 
> I'd favor introducing complexity when it is needed.

That approach does extend the complexity though to migration. When we later want to introduce other first-run behaviours, we'd need to migrate the "first-run" state to a new one.

> > To me, on the other hand, this is just state. I don't see a difference between a file that is created (and can be removed) and a boolean stored in a file somewhere.
> 
> I definitely understand wanting to keep all persistent gui states in a single file. I do think these states should not be modifiable by the user other than potential interactions with the gui itself, so the keys should not be exposed to the user and not be modifiable by `multipass set`.

WFM

> I read up on the `QSettings` documentation and it suggests using it for storing persistent gui states. So I think in the long run, using our `Settings` infrastructure is the right thing to do, but that requires expanding it to be able handle "hidden" keys, etc.

Or just have a separate `State` singleton that will be internal only.

> Given the time crunch of wanting another release out very soon, I don't see that it is possible to implement it in our `Settings` code right now. we are going to have tech debt for this, so I say we use the simplest way to make it work, like what I currently have. Either that, or drop this notification for now and add it in when we have time to implement it correctly.

Yeah, agreed. If we don't want it in `Settings`, that is too much for now.

Reply:
To be clear, I am not proposing a file per flag, but certificates separate from instances separate from metrics, I think is good. It's like single responsibility applied to files. Unfortunately this new flag just has no good company ATM.

Notice that more info in a single file -> more race potential. By using separate files for different purposes, you effectively delegate synchronization to the OS (threads don't collide in disk IO if they use their own file).

Reply:
Oops, I did not see the last message.

I will say I recognize some appeal in having a dedicated entity to route and gate disk IO, as long as private/public is distinguished. But I agree that would go beyond this PR.

Reply:
> certificates separate from instances separate from metrics, I think is good

FWIW I would say at least these three would be "good company" for this incoming flag:
- `multipassd-send-metrics.yaml`
- `network/multipass_subnet`
- `multipass-unique-id`

Reply:
Suppose we merge with those. If the gui starts around the time metrics are being treated, there is a collision. You now need to implement disk IO synchronization across processes, while that was previously done by the OS for free. I fail to see what exact gain would justify the cost.

Of course we could generalize this flag to a launch count and decide that was useful metrics, or make it some sort of setting, but that would be different motivation altogether.

Perhaps a disk IO entity could be useful in cases where there is already a race for the same data. But do we have many such cases? If we don't, aren't we better off just avoiding them? Having entities encapsulate their data in files?

Reply:
One other point. Settings, by definition, are used by multiple entities (setter and settee). That's why they need a coordinating entity (our `Settings` class). 

(Talking about this, I realized the synchronization mechanism is missing: #1221)